misc/sudoku_solver.c

```diff
@@ -1,93 +1,270 @@
-// recursion problem : Sudoku Solver
-/*You are given an incomplete N*N Sudoku and asked to solve it using the
-following recursive algorithm: (1) Scan the Sudoku from left to right row-wise
-to search for an empty cell. (2) If there are no empty cells, print the Sudoku.
-Go to step 5. (3) In the empty cell, try putting numbers 1 to N while ensuring
-that no two numbers in a single row, column, or box are same. Go back to step 1.
-(4) Declare that the Sudoku is Invalid.
-(5) Exit.*/
-
+/**
+ * @file
+ * @brief Sudoku Solver using recursive implementation of brute-force algorithm
+ *
+ * @details
+ * Given an incomplete N*N Sudoku and asked to solve it using the
+ * following recursive algorithm:
+ * 1. Scan the Sudoku from left to right row-wise to search for an empty cell.
+ * 2. If there are no empty cells, print the Sudoku. Go to step 5.
+ * 3. In the empty cell, try putting numbers 1 to N
+ * while ensuring that no two numbers in a single row, column, or box are same.
+ * Go back to step 1.
+ * 4. Declare that the Sudoku is Invalid.
+ * 5. Exit.
+ *
+ * @authors [Anuj Shah](https://github.com/anujms1999)
+ * @authors [Krishna Vedala](https://github.com/kvedala)
+ */
+#include <assert.h>
+#include <inttypes.h>
+#include <math.h>
+#include <stdbool.h>
 #include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
 
-#define M 144
-int N, R, C;
+/** @addtogroup sudoku Sudoku solver
+ * @{
+ */
+/** Structure to hold the matrix and dimensions
+ */
+struct sudoku
+{
+    uint8_t *a; /**< matrix as a flattened 1D row-major array */
+    uint8_t N;  /**< number of elements */
```

Comment:
instead of elements and block of elements, you can use rows and column for better understanding

Reply:
true, then it would probably become a parallel implementation and a slightly different algorithm, wouldn't you say

Reply:
@kvedala Yes but the repository is for learning purpose and if terms like row, columns are used it would be better to understand. So, if you atleast mention rows and columns in brackets in the comment lines along with the previous comments,  it may be of help.

Reply:
I agree and I have fixed much of the 100+files in this repo till now by myself many of them did not even compile and/or execute like this particular file.
In this particular instance, why would we need 3 additional variables?
Moreover, this is where the documentation helps, where each variable is documented for every function and if documented correctly, also provides click-able links to their references wherever they are used within the code.

Reply:
If you rather, I can unstage the changes to the non-executable sudoku file and await for a better pull-request from you.

Reply:
I am saying just modify the /**< number of elements */ to /**< number of elements (columns)*/ and /**< block of elements */ to /**< block of elements (rows)*/

Reply:
ah, ok. done.

Reply:
I dont think you understood the code correctly. 
* `N` is the number of elements = number of rows = number of columns = number of sequential numerical numbers from 1 to N
* `N2` is the block size - which is always `sqrt(N)`

The suggestion doesn't make sense.

Reply:
You can read the details what a sudoku is here: https://en.wikipedia.org/wiki/Mathematics_of_Sudoku#Sudokus_of_other_sizes